src/service/ops/buybackVsp.js

```diff
@@ -0,0 +1,209 @@
+  return Promise.all(promises).then(function ([balance, pricePerShare, withdrawFee]) {
+    const withdrawFeeAmount = balance.mul(withdrawFee).div(DECIMAL18)
+    // all vTokens are 18 decimals
+    return balance.sub(withdrawFeeAmount).mul(pricePerShare).div(DECIMAL18)
+  })
+}
+
+function prepareVspBuyback(data, batchJobPromises) {
+  batchJobPromises.push(data.contract.populateTransaction.swapForVspAndTransferToVVSP(data.asset, data.amount))
+  return prepareJob(data, batchJobPromises)
+}
+
+function prepareUnwrapAndBuyback(data, batchJobPromises) {
+  return getContractMetadata([]).then(async function (pools) {
+    const poolPromises = []
+    pools.forEach(function (poolObj) {
+      poolPromises.push(getPoolContract(poolObj.contract.address, data.wallet.provider).token())
+    })
+    return Promise.all(poolPromises).then(function (tokens) {
+      const vTokenBalPromises = []
+      const matchedVToken = []
+      tokens.forEach(function (token, index) {
+        if (token === data.asset) {
+          const vTokenContract = new ethers.Contract(token, vesperERC20TokenAbi, data.wallet.provider)
+          matchedVToken.push(pools[index])
+          vTokenBalPromises.push(vTokenContract.balanceOf(revenueConfig.buybackAddress))
+        }
+      })
+
+      // Prepare mapping of vToken and it's balance.
+      const vTokenBalMapping = {}
+      return Promise.all(vTokenBalPromises).then(function (vTokenBalances) {
+        vTokenBalances.forEach(function (vTokenBalance, matchVTokenIndex) {
+          const key = matchedVToken[matchVTokenIndex].contract.address
+          vTokenBalMapping[key] = vTokenBalance
+        })
+
+        // sort vToken's balance in decreasing order
+        const vTokenBalMappingSorted = Object.fromEntries(Object.entries(vTokenBalMapping).sort((a, b) => b[1] - a[1]))
+        const vTokens = Object.keys(vTokenBalMappingSorted)
+
+        // Calculate vToken value in collateral token
+        const balancePromise = []
+        vTokens.forEach(function (vToken) {
+          const poolContract = getPoolContract(vToken, data.wallet.provider)
+          balancePromise.push(getBalanceAmountInCollateralToken(poolContract))
+        })
+
+        let availableBalance = data.buybackBalance
+        const requiredAmount = data.amount
+        let index = 0
+        return Promise.all(balancePromise).then(function (balances) {
+          // iterate and unwrap till we have enough collateral token balance for buyback
+          while (requiredAmount.gt(availableBalance) && index < balances.length) {
+            // prepare unwrap job
+            batchJobPromises.push(data.contract.populateTransaction.unwrapAll(vTokens[index]))
+            availableBalance = availableBalance.add(balances[index])
+            index++
+          }
+          // Buy VSP only if enough balance.
+          if (availableBalance.gte(requiredAmount)) {
+            return prepareVspBuyback(data, batchJobPromises)
```

Comment:
Seems that this function will unwrap all `vTokens` with `balance > requiredAmount` but swap for `VSP` only the required asset (data.asset) is that right?

Reply:
@marcelomorgado, This will unwrap only one vToken but with the full amount (whatever is available in buyback contract for that vToken).
Yes, it will swap for only one asset. 

This is the flow
- Input : collateral token address and buyback amount 
- if buyback amount >= collateral token balance at buyback contract, unwrap is not needed and just do buyback
-  if buyback amount < collateral token balance at buyback contract, we need to unwrap first
    - Scan all pools and find out which pool is using provided collateral token. Multiple pools can use the same collateral token.
    - Convert vToken balance into collateral token balance
    - Sort and find out which one provides max collateral token value
    - Unwrap till we have buyback value.
    - Lastly do buyback.
    - Single batch txn is submitted for unwrap and buyback.